### PR TITLE
feat(new sink): Initial `rabbitmq` sink implementation

### PR DESCRIPTION
**Pull request by AlyHKafoury**
Signed-off-by: AlyHKafoury <aly.kafoury@gmail.com>

### CODE REVIEW

**Comment by AlyHKafoury** [comment]
@LucioFranco I've created this as a draft and for you to follow the progress from early stages if it's ok to guide me early for a better path

File: src/sinks/rabbitmq.rs [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+  type SinkItem = Event;
+  type SinkError = ();
+
+  fn start_send(&mut self, item: Self::SinkItem) -> StartSend<Self::SinkItem, Self::SinkError> {
+    let payload = encode_event(&item, &self.encoding);
+    let future = self.channel.basic_publish(
+      "",
+      &self.queue_name,
+      payload,
+      BasicPublishOptions::default(),
```
Comment (LucioFranco) [inline src/sinks/rabbitmq.rs]: we may want to expose these options?

File: src/sinks/rabbitmq.rs [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+      BasicProperties::default(),
+    );
+    self.in_flight.push(future.join(future::ok(())));
```
Comment (LucioFranco) [inline src/sinks/rabbitmq.rs]: you probably want to join the ack id here so that you can ack it below in the ready match arm.

**Comment by AlyHKafoury** [comment]
@LucioFranco the acker in the integration test gets out a strange number:

thread 'sinks::rabbitmq::integration_test::publish_messages' panicked at 'assertion failed: `(left == right)`
  left: `499500`,
 right: `1000`', src/sinks/rabbitmq.rs:329:5

I don't understand where this value comes from as the sequence number is correct in the sync

**Comment by LucioFranco** [comment]
@AlyHKafoury so I think there might be something wrong when you decide to ack. It seems that you may be calling https://github.com/timberio/vector/blob/master/src/buffers/mod.rs#L143 more than you think you are? I would assume the code where you call ack just might be getting called a lot more than expected?

**Comment by AlyHKafoury** [comment]
> @AlyHKafoury were you able to figure it out?

@LucioFranco yea, I was acking the sequence number of the messages instead the count of the messages that was the wrong issue the number 499500 is the summation of all the numbers from 0 to 999 :D :D !!

**Comment by LucioFranco** [comment]
@AlyHKafoury so I think the code mostly looks good! I think the next step is to add a docker container that we can run the tests aginst. We have a `docker` feature flag you can add your feature flag to that. Then we just need to add that docker contianer to `docker-compose` and to `circleci` test-stable job.

Let me know if you have any trouble, I am happy to help!

**Comment by AlyHKafoury** [comment]
@LucioFranco Can you please check the latest commit is there something I am missing yet ?

**Comment by binarylogic** [comment]
@Jeffail tagging you since you have experience with RabbitMQ. 😄

**Comment by Jeffail** [comment]
Hey @AlyHKafoury, I think it might be worth refactoring the configuration parameters here slightly. 

When defining an AMQP sink I would only expect to specify an exchange and a routing key. It's often the case that the routing key will be the same as the target queue, but sometimes the queue bindings will be defined separately and only known to consumers.

I would therefore not expect to see any queue based fields in a sink config, including the ability to declare a queue. However, I would want to be able to declare the target exchange, but this should be optional.

**Comment by binarylogic** [comment]
@AlyHKafoury thanks again for doing this. Do you plan on making the suggested changes? We'd really appreciate it. If not, we can assign someone to finish this off.

**Comment by AlyHKafoury** [comment]
@binarylogic Sure I will I just missed the comments.

**Comment by binarylogic** [comment]
@AlyHKafoury thanks, and no problem if you can't finish this off. We're planning to have @Jeffail work on it this week, so feel free to chime if you plan to do it yourself.

**Comment by Jeffail** [comment]
Hey @AlyHKafoury, got the changes we need on this branch: https://github.com/timberio/vector/tree/AlyHKafoury-rabbitmq-sink, I'm just finishing up (adding some extra integration tests) and then I'll get it merged.

**Comment by Jeffail** [comment]
Superseded by: https://github.com/timberio/vector/pull/1376

**Pull request closed without merging**